### ISSUE TITLE
Batch splitting for distributed training

### ISSUE THREAD

**Issue by tboo** [uber/ludwig#151]
I am performing image classification (with success :)) but would like to improve prediction accuracy. Since this is related to batch size and my images are rather large (500x250) I am constrained by the GPU memory. I have successfully distributed the training on a GPU cluster using horovod and openMPI according to the documentation with the `stacked_cnn` architecture w/o modifications. From what it looks like (to me) the batches themselves are not split but the batches are distributed in their entity. Because when I try to increase the batch size I get errors like that which disappear when I lower the batch size again. 

```
ResourceExhaustedError (see above for traceback): OOM when allocating tensor with shape[32,256,250,500] and type float on /job:localhost/replica:0/task:0/device:GPU:0 by allocator GPU_0_bfc
```

My question is: is batch splitting across multiple GPU nodes possible at all?

**Comment by w4nderlust** [comment]
Let me explain a little bit how things work, i may in the future copy this explanation inside the docs:
When you use `--horovod` what happens is that every worker loads the entire data in memory (you can specify to read it form the hdf5 file if it doesn't fit in memory). Each worker will then run a different batch in parallel of `batch_size`. The fact the the batches are different is ensured by the `DistributedBatcher` class.
This means that if you specify batch 50 and you have 8 workers, you are effectively working with a batch of 400. The horovod implementation compensates for this adjusting the learning rate. You can look at the itnernals if you are interested.
Now, directly to your problem: 500x250 images are huge for current standards in available GPU memory, and that's the reason you are getting the `ResourceExhaustedError`. Depending on the task, downsampling them to much smaller dimensions should work really well, not decreasing performances a lot (in most cases improving them), reducing ram consumption, speeding up computation, and allowing you to use the freed ram with a larger batch size, that will speed up computation even further.
The way to do it is providing width and heigh parameters to the image preprocessing (if you do, preprocessing needs to be done again, make sure you delete the hdf5 and json files). At the moment the parameters should be set in the image feature itself, like `{name: img, type: image, width: 128, height: 128}`, but there is an open pull request to fix this to `{name: img, type: image, preprocessing: {width: 128, height: 128}}` which will be accepted most likely by the end of the week.
Let me know if this is clear and if applying this suggestions you are able to solve your issues.

**Comment by tboo** [comment]
Thanks so much for taking the time to clarify the data processing to a non-computer scientist :) I do realize that the image dimensions are in my case far larger than what people usually use in image classification. The problem is: 500x250 is already after 8 fold down-sampling of the original images. Since these are biological images you reach at some point a sub-sampling sweet spot at which you get rid of the distinguishing details. Nevertheless I will have to get a bit creative (crops) but it helps to understand how the data are handled. So thanks again and keep on the awesome work.

**Comment by w4nderlust** [comment]
Got it, yeas I can imagine how in some usecases resolution is really important. I would still suggest you to try downsizing through preprocessing as a trade-off, some pattern that lead to some outcomes may be missed, but maybe the accuracy that you get is good enough. It all depends on your task and expected performance.
Another possible trade-off is model size, for instance the number for filters for each layer and the number of layers. If you decrease the number of filters at the first layers (the ones that work on the full size fo the image) and increase them ad the last layers, you can potentially use larger images without sacrificing performance. For instance [VGG](https://gist.github.com/baraldilorenzo/07d7802847aaad0a35d3) has first 64 filter for the first 2 layers, than 128 for the following 2 layers, then 256 and then 516. You can do something similar with the first 4 layers for instance with just 8 / 16 / 32 filters.
